### ISSUE TITLE
Preprocessing in multilingual translation: Dictionary creation on multiple language pairs

### ISSUE THREAD

**Issue by villmow** [pytorch/fairseq#1755]
## ❓ Questions and Help

I try to understand the preprocessing pipeline of multilingual translation as explained in the [translation README](https://github.com/pytorch/fairseq/blob/master/examples/translation/README.md):

The following code states how we should preprocess the data for multiple language pairs:

 1. First binarize the first language pair and create a joined dictionary:
```bash
# Binarize the de-en dataset
TEXT=examples/translation/iwslt17.de_fr.en.bpe16k
fairseq-preprocess --source-lang de --target-lang en \
    --trainpref $TEXT/train.bpe.de-en --validpref $TEXT/valid.bpe.de-en \
    --joined-dictionary \
    --destdir data-bin/iwslt17.de_fr.en.bpe16k \
    --workers 10
```
 2. Then re-use that dictionary during binarization of the second language pair:
```bash
# Binarize the fr-en dataset
# NOTE: it's important to reuse the en dictionary from the previous step
fairseq-preprocess --source-lang fr --target-lang en \
    --trainpref $TEXT/train.bpe.fr-en --validpref $TEXT/valid.bpe.fr-en \
    --joined-dictionary --tgtdict data-bin/iwslt17.de_fr.en.bpe16k/dict.en.txt \
    --destdir data-bin/iwslt17.de_fr.en.bpe16k \
    --workers 10
``` 

**Question**: Doesn't this lead to words, which occur only in the second language pair but not in the first, being replaced with unknown? The dictionary is created only on the first language pair. Will new words (from the training set of the second language pair) be added to the dictionary, or are they ignored/replaced with unknown?

**Comment by myleott** [comment]
Sort of. We only reuse the target dictionary, in this case "en". So any English words that only appear in the second dataset will be replaced by unknown. But in the end you should have three dictionaries: en, de, fr.

**Comment by myleott** [comment]
Thanks, good catch. You should remove the `--joined-dictionary` option, that way it reuses the English dictionary, but will generate a new French dictionary.

I'll update the instructions.

**Comment by myleott** [comment]
Fixed in e5facc9984bcfe3de8486db2639cd3c228d66dfa

**Comment by ishanshrivastava2011** [comment]
Hi @myleott 
Preprocessing it this way, works when the model architecture does not share an encoder. I am trying to train an architecture with a shared encoder for the same Multilingual Translation task. When I follow the steps in the example and train on this architecture (by using the  --share-encoders flag), I get the following error:

`ValueError: --share-*-embeddings requires a joined dictionary: --share-encoder-embeddings requires a joined source dictionary, --share-decoder-embeddings requires a joined target dictionary, and --share-all-embeddings requires a joint source + target dictionary.`

So, I believe I would need to create a joined dictionary but based on your reply above using --joined-dictionary won't work as I would not get a new french dictionary. 

Could you please suggest how can I achieve a joined dictionary such that I can train an architecture with a shared encoder for the same Multilingual Translation task?
